pkg/cloud/services/eks/addons.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/eks"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	ekscontrolplanev1 "sigs.k8s.io/cluster-api-provider-aws/controlplane/eks/api/v1alpha3"
+	eksaddons "sigs.k8s.io/cluster-api-provider-aws/pkg/eks/addons"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/record"
+)
+
+func (s *Service) reconcileAddons(ctx context.Context) error {
+	s.scope.Info("Reconciling EKS addons")
+
+	eksClusterName := s.scope.KubernetesClusterName()
+
+	// 1. get addons from AWS for cluster
+	s.scope.V(2).Info("getting installed eks addons", "cluster", eksClusterName)
+	installed, err := s.getClusterAddons()
+	if err != nil {
+		return fmt.Errorf("getting installed eks addons: %w", err)
+	}
+
+	// 2. get addons from spec
+	desiredAddons := s.translateToSdkAddon(s.scope.Addons())
+
+	// 3. Exit early if there is none
+	if len(installed) == 0 && len(desiredAddons) == 0 {
+		s.scope.Info("no addons installed and no addons to install, no action needed")
+		return nil
+	}
+
+	// 3. compute CUD operations between 1 & 2
+	s.scope.V(2).Info("creating eks addons plan", "cluster", eksClusterName, "numdesired", len(desiredAddons), "numinstalled", len(installed))
+	addonsPlan := eksaddons.NewAddonsPlan(eksClusterName, desiredAddons, installed, s.EKSClient)
+	procedures, err := addonsPlan.Create(ctx)
+	if err != nil {
+		s.scope.Error(err, "failed creating eks addons plane")
+		return fmt.Errorf("creating eks addons plan: %w", err)
+	}
+	s.scope.V(2).Info("computed EKS addons plan", "numprocs", len(procedures))
+	if len(procedures) == 0 {
+		s.scope.Info("addon plan has no actions required, no action needed")
+		return nil
+	}
+
+	// 4. Perform required operations
+	for _, procedure := range procedures {
+		s.scope.V(2).Info("Executing addon procedure", "name", procedure.Name())
+		if err := procedure.Do(ctx); err != nil {
+			s.scope.Error(err, "failed executing addon procedure", "name", procedure.Name())
+			return fmt.Errorf("%s: %w", procedure.Name(), err)
+		}
+	}
+
+	// 5. Update status with addons installed details
+	// NOTE: we should be able to construct this froom the plan/procedures????
```

Comment:
Could we add a `getAllDesired()` to `addonsPlan` so we don't need to make the extra API calls here?

Reply:
I did wonder whether we should construct the Status from the operations. I was undecided and also I wanted to leave `Plan\Procedure` generic. But I guess I can return `AddonsPlan` from the `New` instead of `Plan` and then have it on the `AddonsPlan`  instead. I'll make that change....thanks for the feedback on this

Reply:
Just looked into it a bit more and it doesn't look like we're capturing the ARN during the plan/procedure bit.

Reply:
In the `CreateAddonProcedure` i can get that from the output and save it.

Reply:
I decided to go with querying the addons again for the scenario where there are no actions to take with the plan but where there might be an update to the addon...for example the status moving from UPDATING to ACTIVE.